src/blog/understanding-ecmascript-part-4.md

```diff
@@ -0,0 +1,280 @@
+---
+title: 'Understanding the ECMAScript spec, part 4'
+author: '[Marja Hölttä](https://twitter.com/marjakh), speculative specification spectator'
+avatars:
+  - marja-holtta
+date: 2020-05-06
+tags:
+  - ECMAScript
+description: 'Tutorial on reading the ECMAScript specification'
+tweet: ''
+---
+
+## Previous episodes
+
+In [part 1](/blog/understanding-ecmascript-part-1), we read through a simple method — `Object.prototype.hasOwnProperty` — and the **abstract operations** it invokes. We familiarized ourselves with the shorthands `?` and `!` related to error handling. We encountered **language types**, **specification types**, **internal slots**, and **internal methods**.
+
+In [part 2](/blog/understanding-ecmascript-part-2), we examined a simple grammar production and how its runtime semantics are defined. In [the extra content](/blog/extra/understanding-ecmascript-part-2-extra), we also followed a long grammar production chain from `AssignmentExpression` to `MemberExpression`.
+
+In [part 3](/blog/understanding-ecmascript-part-3), we familiarized ourselves with the lexical grammar, the syntactic grammar, and the shorthands used for defining the syntactic grammar.
+
+## Meanwhile in other parts of the Web
+
+[Jason Orendorff](https://github.com/jorendorff) from Mozilla published [a great in-depth analysis of JS syntactic quirks](https://github.com/mozilla-spidermonkey/jsparagus/blob/master/js-quirks.md#readme). Even though the implementation details differ, every JS engine faces the same problems with these quirks.
+
+## Cover grammars
+
+In this episode, we take a deeper look into *cover grammars*. They are a way to specify grammar rules for syntactic constructs where we don't know what we're looking at until we've seen the complete construct.
+
+Again, we'll skip the subscripts for `[In, Yield, Await]` for brevity, as they aren't important for this blog post. See [part 3](/blog/understanding-ecmascript-part-3) for an explanation of their meaning and usage.
+
+## Parenthesized expression or an arrow parameter list?
+
+Typically, parsers decide which grammar production to follow based on finite lookahead.
+
+For example:
+```javascript
+let x = (a,
+```
+
+Is this the start of an arrow function, like this?
+
+```javascript
+let x = (a, b) => { return a + b };
+```
+
+Or maybe it's a parenthesized expression, like this?
+
+```javascript
+let x = (a, 3);
+```
+
+The parenthesized whatever-it-is can be arbitrarily long - we cannot know what it is based on a finite amount of tokens.
+
+If the productions were written like this:
+
+```grammar
+AssignmentExpression :
+...
+ConditionalExpression (eventually leading to PrimaryExpression)
```

Comment:
Can you please check how this renders locally? I suspect `(...)` will be treated as part of the syntax.

Reply:
Also, here and elsewhere the production variants should be indented with 2 spaces (see part 3 Markdown for an example).

Reply:
Added spaces, also took the "comment inside grammar" out of the grammar element, so now we have 2 grammar elements here, but prob that's the best we can do.